File: homework/src/pa1/List.java
package pa1;

import pa1.library.IIntListADT;

public class List implements IIntListADT {
	
	private static final int CURSOR_INDEX_INVALID = -1;
	
	private Node<Integer> cursor;
	private Node<Integer> front;
	private Node<Integer> back;
	private int cursorIndex = CURSOR_INDEX_INVALID;
	private int length;
	
	public List() { 
		
	}
	
	/* Required Methods */

	@Override
	public int length() {
		return this.length;
	}

	@Override
	public int index() {
		if (isNodeDefined(this.cursor)) {
			return cursorIndex;
		}
		return -1;
	}

	@Override
	public int front() {
		return this.front.get().intValue();
	}

	@Override
	public int back() {
		return this.back.get().intValue();
	}

	@Override
	public int get() {
		return this.cursor.get().intValue();
	}

	@Override
	public boolean equals(IIntListADT passedList) {
		if (this.length() == passedList.length()) {
			if (!this.isEmpty()) {
				Node<Integer> thisListIterator = this.getFrontNode();
				Node<Integer> passedListIterator = this.getFrontNode();
				while (this.isNodeDefined(thisListIterator)) {
					if (!thisListIterator.get().equals(passedListIterator.get())) {
						return false;
					}
					thisListIterator = thisListIterator.getNext();
					passedListIterator = passedListIterator.getNext();
				}
				return true;
			}
		}
		return false;
	}

	@Override
	public void clear() {
		if (!this.isEmpty()) {
			Node<Integer> thisListIterator = this.getFrontNode();
			while (this.isNodeDefined(thisListIterator)) {
				thisListIterator = thisListIterator.getNext();
				thisListIterator.previous.reset();
			}
		}
		this.front = null;
		this.back = null;
		this.cursor = null;
		this.cursorIndex = CURSOR_INDEX_INVALID;
		this.length = 0;
	}

	@Override
	public void moveFront() {
		if (isNodeDefined(this.front)) {
			this.cursor = this.front;
			this.cursorIndex = 0;
		}
	}

	@Override
	public void moveBack() {
		if (isNodeDefined(this.back)) {
			this.cursor = this.back;
			this.cursorIndex = (this.length - 1);
		}
	}

	@Override
	public void movePrev() {
		this.cursor = this.cursor.getPrevious();
		if (!isNodeDefined(this.cursor)) {
			this.cursorIndex = CURSOR_INDEX_INVALID;
		}
		else {
			this.cursorIndex -= 1;
		}
	}

	@Override
	public void moveNext() {
		this.cursor = this.cursor.getNext();
		if (!isNodeDefined(this.cursor)) {
			this.cursorIndex = CURSOR_INDEX_INVALID;
		}
		else {
			this.cursorIndex += 1;
		}
	}

	@Override
	public void prepend(int passedData) {
		Node<Integer> newNode = this.newNode(passedData);
		if (isNodeDefined(this.front)) {
			newNode.setNext(this.front);
			this.front.setPrevious(newNode);
		}
		this.front = newNode;
		this.length += 1;
		this.cursorIndex += 1;
	}

	@Override
	public void append(int passedData) {
		Node<Integer> newNode = this.newNode(passedData);
		if (isNodeDefined(this.back)) {
			newNode.setPrevious(this.back);
			this.back.setNext(newNode);
		}
		this.back = newNode;
		this.length += 1;
	}

	@Override
	public void insertBefore(int passedData) {
		Node<Integer> newNode = this.newNode(passedData);
		newNode.setNext(this.cursor);
		newNode.setPrevious(this.cursor.getPrevious());
		this.cursor.setPrevious(newNode);
		this.length += 1;
		this.cursorIndex += 1;
	}

	@Override
	public void insertAfter(int passedData) {
		Node<Integer> newNode = this.newNode(passedData);
		newNode.setNext(this.cursor.getNext());
		newNode.setPrevious(this.cursor);
		this.cursor.setNext(newNode);
		this.length += 1;
	}

	@Override
	public void deleteFront() {
		if (this.cursor.equals(this.front)) {
			this.cursorIndex = CURSOR_INDEX_INVALID;
		}
		this.removeNode(this.front);
	}

	@Override
	public void deleteBack() {
		if (this.cursor.equals(this.back)) {
			this.cursorIndex = CURSOR_INDEX_INVALID;
		}
		this.removeNode(this.back);
	}

	@Override
	public void delete() {
		this.removeNode(this.cursor);
	}

	@Override
	public IIntListADT copy() {
		List newList = new List();
		Node<Integer> thisListIterator = this.getFrontNode();
		while (this.isNodeDefined(thisListIterator)) {
			newList.append(thisListIterator.get().intValue());
			thisListIterator = thisListIterator.getNext();
		}
		return newList;
	}

	@Override
	public IIntListADT concat(IIntListADT passedList) {
		List newList = new List();
		Node<Integer> thisListIterator = this.getFrontNode();
		while (this.isNodeDefined(thisListIterator)) {
			newList.append(thisListIterator.get().intValue());
			thisListIterator = thisListIterator.getNext();
		}
		return null;
	}
	
	public String toString() {
		Node<Integer> thisListIterator = this.getFrontNode();
		String stringRepresentation = "";
		while (this.isNodeDefined(thisListIterator)) {
			stringRepresentation += thisListIterator.toString();
			thisListIterator = thisListIterator.getNext();
			if (thisListIterator != null) {
				stringRepresentation += " ";
			}
		}
		return stringRepresentation;
	}
	
	public Node<Integer> getFrontNode() {
		return this.front;
	}
	
	public Node<Integer> getBackNode() {
		return this.back;
	}
	
	public Node<Integer> getCursorNode() {
		return this.cursor;
	}
	
	public boolean isEmpty() {
		return (this.length > 0);
	}
	
	public boolean isNodeDefined(Node<?> passedNode) {
		return (passedNode != null) && (passedNode.get() != null) && (passedNode.getOwner().equals(this));
	}
	
	private Node<Integer> newNode(int passedValue) {
		Node<Integer> newNode = new Node<Integer>(this, new Integer(passedValue));
		return newNode;
	}
	
	private void removeNode(Node<Integer> passedNode) {
		if (isNodeDefined(passedNode)) {
			Node<Integer> nextNode = passedNode.getNext();
			Node<Integer> previousNode = passedNode.getPrevious();
			if (isNodeDefined(nextNode)) {
				nextNode.setPrevious(passedNode.getPrevious());
			}
			if (isNodeDefined(previousNode)) {
				previousNode.setNext(passedNode.getNext());
			}
			passedNode.reset();
			this.length -= 1;
		}
	}
	
	/* Node Implementation */
	
	private class Node<T> {
		private final List owningList;
		private T element;
		private Node<T> previous;
		private Node<T> next;
		
		public Node(List passedOwningList, T passedElement) {
			this.owningList = passedOwningList;
			this.element = passedElement;
		}
		
		public T get() {
			return this.element;
		}
		
		public List getOwner() {
			return this.owningList;
		}
		
		public void setValue(T passedValue) {
			this.element = passedValue;
		}
		
		public void reset() {
			this.element = null;
			this.previous = null;
			this.next = null;
		}
		
		public Node<T> getNext() {
			return this.next;
		}
		
		public Node<T> getPrevious() {
			return this.previous;
		}
		
		public void setNext(Node<T> passedNode) {
			this.next = passedNode;
		}
		
		public void setPrevious(Node<T> passedNode) {
			this.previous = passedNode;
		}
		
		public String toString() {
			return this.element.toString();
		}
	}

}

File: pa2/src/List.c
/*********************************************************************
 * Malcolm Riley
 *
 * CruzID: masriley
 * Assignment: pa2
 *
 * 10-2017
 *********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include "List.h"

/* Internal Function Declarations */
void exitBadWithMessage(const char* passedCharArray);
void insertIntoEmpty(List passedList, Node passedNode);
int insertNodeBefore(List passedList, Node passedNode, Node passedInsertedNode);
int insertNodeAfter(List passedList, Node passedNode, Node passedInsertedNode);
int removeNode(List passedList, Node passedNode);
int isNull(void* passedPointer, char* passedCharArray, int passedIsTerminal);
int isListEmpty(List passedList, char* passedCharArray, int passedIsTerminal);
int validateCursor(List passedList);
int isCursorValid(List passedList, char* passedCharArray, int passedIsTerminal);

Node newNode(int passedValue) {
	Node newNode = malloc(sizeof(NodeObject));
	newNode->nextNode = NULL;
	newNode->previousNode = NULL;
	newNode->value = passedValue;
	return newNode;
}

void freeNode(Node* passedNode) {
	if ((passedNode != NULL) && (*passedNode != NULL)) {
		free(*passedNode);
		*passedNode = NULL;
	}
}

/* Node Functions */

int nodesAreEqual(Node passedFirstNode, Node passedSecondNode) {
	if ((passedFirstNode != NULL) && (passedSecondNode != NULL)) {
		return (passedFirstNode->value == passedSecondNode->value);
	}
	return FALSE;
}

/* List Constructor/Destructor */
List newList(void) {
	List newList = malloc(sizeof(ListObject));
	newList->cursorIndex = UNDEFINED;
	newList->length = 0;
	newList->nodeBack = NULL;
	newList->nodeFront = NULL;
	newList->nodeCursor = NULL;
	return newList;
}

void freeList(List* passedList) {
	if (!isNull(&passedList, "Error while freeing List:", FALSE)) {
		while(length(*passedList) > 0) {
			deleteBack(*passedList);
		}
	}
	free(passedList);
	passedList = NULL;
}

/* List Functions */
int length(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the length of a null List.", TRUE)) {
		return passedList->length;
	}
	return UNDEFINED;
}

int index(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the index of a null List.", TRUE)) {
		validateCursor(passedList);
		return passedList->cursorIndex;
	}
	return UNDEFINED;
}

int front(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the front of a null List.", TRUE)) {
		if (!isNull(passedList->nodeFront, "Front Node is NULL.", FALSE)) {
			return passedList->nodeFront->value;
		}
	}
	return UNDEFINED;
}


int back(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the back of a null List.", TRUE)) {
		if (!isNull(passedList->nodeBack, "Back Node is NULL.", FALSE)) {
			return passedList->nodeBack->value;
		}
	}
	return UNDEFINED;
}

int get(List passedList) {
	if (!isNull(passedList, "Cannot retrieve the cursor of a null List.", TRUE)) {
		if (!isNull(passedList->nodeCursor, "Cursor Node is NULL.", FALSE)) {
			return passedList->nodeCursor->value;
		}
	}
	return UNDEFINED;
}

int equals(List passedFirstList, List passedSecondList) {
	if ((!isNull(passedFirstList, "Equals: First list is null.", FALSE)) && (!isNull(passedSecondList, "Equals: Second list is null.", FALSE))) {
		int areEqual = (passedFirstList->length == passedSecondList->length);
		if (areEqual) {
			Node firstNode = passedFirstList->nodeFront;
			Node secondNode = passedSecondList->nodeFront;
			for(int ii = 0; (ii < passedFirstList->length) && (areEqual); ii += 1) {
				areEqual = nodesAreEqual(firstNode, secondNode);
				firstNode = firstNode->nextNode;
				secondNode = secondNode->nextNode;
			}
		}
		return areEqual;
	}
	return FALSE;
}

void clear(List passedList) {
	if (!isListEmpty(passedList, "Error with List when clear was called:", FALSE)) {
		while (passedList->length > 0) {
			deleteBack(passedList);
		}
	}
}

void moveFront(List passedList) {
	if (!isListEmpty(passedList, "Error with List when moveFront was called:", TRUE)) {
		if (!isNull(passedList->nodeFront, "Error with List's front Node:", FALSE)) {
			passedList->nodeCursor = passedList->nodeFront;
			passedList->cursorIndex = 0;
		}
	}
}

void moveBack(List passedList) {
	if (!isListEmpty(passedList, "Error with List when moveBack was called:", TRUE)) {
		if (!isNull(passedList->nodeBack, "Error with List's back Node:", FALSE)) {
			passedList->nodeCursor = passedList->nodeBack;
			passedList->cursorIndex = (length(passedList) - 1);
		}
	}
}

void movePrev(List passedList) {
	if (!isListEmpty(passedList, "Error with List when movePrev was called:", FALSE)) {
		if (!isNull(passedList->nodeCursor, "Error with List's Cursor Node:", FALSE)) {
			passedList->nodeCursor = passedList->nodeCursor->previousNode;
			passedList->cursorIndex -= 1;
		}
	}
}

void moveNext(List passedList) {
	if (!isListEmpty(passedList, "Error with List when moveNext was called:", FALSE)) {
		if (!isNull(passedList->nodeCursor, "Error with List's Cursor Node:", FALSE)) {
			passedList->nodeCursor = passedList->nodeCursor->nextNode;
			passedList->cursorIndex += 1;
		}
	}
}

void prepend(List passedList, int passedValue) {
	if (!isNull(passedList, "Cannot prepend to a null List!", TRUE)) {
		Node allocatedNode = newNode(passedValue);
		if (passedList->length < 1) {
			insertIntoEmpty(passedList, allocatedNode);
		}
		else if (insertNodeBefore(passedList, passedList->nodeFront, allocatedNode)) {
			passedList->cursorIndex += 1;
			passedList->length += 1;
			passedList->nodeFront = allocatedNode;
		}
		else {
			freeNode(&allocatedNode);
		}
	}
}

void append(List passedList, int passedValue) {
	if (!isNull(passedList, "Cannot append to a null List!", TRUE)) {
		Node allocatedNode = newNode(passedValue);
		if (passedList->length < 1) {
			insertIntoEmpty(passedList, allocatedNode);
		}
		else if (insertNodeAfter(passedList, passedList->nodeBack, allocatedNode)) {
			passedList->length += 1;
			passedList->nodeBack = allocatedNode;
		}
		else {
			freeNode(&allocatedNode);
		}
	}
}

void insertBefore(List passedList, int passedValue) {
	if (!isNull(passedList, "Cannot insert into a null List!", TRUE)) {
		if (isCursorValid(passedList, "Cannot insert before an undefined cursor!", TRUE)) {
			Node allocatedNode = newNode(passedValue);
			if (insertNodeBefore(passedList, passedList->nodeCursor, allocatedNode)) {
				passedList->cursorIndex += 1;
				passedList->length += 1;
			}
			else {
				freeNode(&allocatedNode);
			}
		}
	}
}

void insertAfter(List passedList, int passedValue) {
	if (!isNull(passedList, "Cannot insert into a null List!", TRUE)) {
		if (isCursorValid(passedList, "Cannot insert after an undefined cursor!", TRUE)) {
			Node allocatedNode = newNode(passedValue);
			if (insertNodeAfter(passedList, passedList->nodeCursor, allocatedNode)) {
				passedList->length += 1;
			}
			else {
				freeNode(&allocatedNode);
			}
		}
	}
}

void deleteFront(List passedList) {
	if (!isListEmpty(passedList, "Error when deleting the front Node of a List", TRUE)) {
		if (removeNode(passedList, passedList->nodeFront)) {
			passedList->cursorIndex -= 1;
			passedList->length -= 1;
		}
	}
}

void deleteBack(List passedList) {
	if (!isListEmpty(passedList, "Error when deleting the back Node of a List", TRUE)) {
		if (removeNode(passedList, passedList->nodeBack)) {
			passedList->length -= 1;
		}
	}
}

void delete(List passedList) {
	if (!isListEmpty(passedList, "Error when deleting the cursor Node of a List", TRUE)) {
		if (removeNode(passedList, passedList->nodeCursor)) {
			passedList->length -= 1;
			passedList->cursorIndex = UNDEFINED;
		}
	}
}

List concatList(List passedFirstList, List passedSecondList) {
	int firstListValid = !isListEmpty(passedFirstList, "Error with prepended List during concatenation:", FALSE);
	int secondListValid = !isListEmpty(passedFirstList, "Error with appended List during concatenation:", FALSE);
	List allocatedList = newList();
	// If both Lists are valid, concatenate
	if ((firstListValid != FALSE) && (secondListValid != FALSE)) {
		List allocatedList = copyList(passedFirstList);
		Node iteratedNode = passedSecondList->nodeFront;
		while (iteratedNode != NULL) {
			append(allocatedList, iteratedNode->value);
			iteratedNode = iteratedNode->nextNode;
		}
		return allocatedList;
	}
	// If only first List is valid, use that one
	else if (firstListValid != FALSE) {
		return copyList(passedFirstList);
	}

	// If only second List is valid, use that one
	else if (secondListValid != FALSE) {
		return copyList(passedSecondList);
	}

	// Otherwise, return empty List
	return allocatedList;
}

void printList(FILE* passedOutputFile, List passedList) {
	if(!isNull(passedOutputFile, "Cannot print to a null file reference.", FALSE)) {
		if (!isListEmpty(passedList, "Error with List during print:", FALSE)) {
			Node iteratedNode = passedList->nodeFront;
			while (iteratedNode != NULL) {
				fprintf(passedOutputFile, "%d", iteratedNode->value);
				iteratedNode = iteratedNode->nextNode;
				if (iteratedNode != NULL) {
					fprintf(passedOutputFile, "%s", " ");
				}
			}
		}
	}
}

List copyList(List passedList) {
	List allocatedList = newList();
	if (!isListEmpty(passedList, "Cannot copy a null List.", FALSE)) {
		Node iteratedNode = passedList->nodeFront;
		while (iteratedNode != NULL) {
			append(allocatedList, iteratedNode->value);
			iteratedNode = iteratedNode->nextNode;
		}
	}
	return allocatedList;
}

/* Internal Functions */

/**
 * If the passedList is empty, inserts into it, setting the front and back nodes.
 */
void insertIntoEmpty(List passedList, Node passedNode) {
	passedList->nodeFront = passedNode;
	passedList->nodeBack = passedNode;
	passedList->length += 1;
}

/**
 * Removes a Node from passedList, freeing it.
 */
int removeNode(List passedList, Node passedNode) {
	if (!isNull(passedList, "Cannot remove a Node from a List that is null.", FALSE)) {
		if (!isNull(passedNode, "Cannot remove a Node that is null.", FALSE)) {
			Node nextNode = passedNode->nextNode;
			Node previousNode = passedNode->previousNode;
			if (nextNode != NULL) {
				nextNode->previousNode = passedNode->previousNode;
				if (passedNode == passedList->nodeFront) {
					passedList->nodeFront = nextNode;
				}
			}
			if (previousNode != NULL) {
				previousNode->nextNode = passedNode->nextNode;
				if (passedNode == passedList->nodeBack) {
					passedList->nodeBack = previousNode;
				}
			}
			freeNode(&passedNode);
			return TRUE;
		}
	}
	return FALSE;
}

/**
 * Inserts passedInsertedNode before passedNode.
 */
int insertNodeBefore(List passedList, Node passedNode, Node passedInsertedNode) {
	int passedNodeValid = !isNull(passedNode, "Attempting to insert before null Node", FALSE);
	int insertedNodeValid = !isNull(passedNode, "Attempting to insert null Node", FALSE);
	if ((passedNodeValid) && (insertedNodeValid)) {
		// Set existing links
		Node previous = passedNode->previousNode;
		if (previous != NULL) {
			previous->nextNode = passedInsertedNode;
		}
		passedNode->previousNode = passedInsertedNode;

		// Set new links
		passedInsertedNode->nextNode = passedNode;
		passedInsertedNode->previousNode = previous;

		// Set new front if applicable
		if (passedList->nodeFront == passedNode) {
			passedList->nodeFront = passedInsertedNode;
		}
		return TRUE;
	}
	return FALSE;
}

/**
 * Inserts passedInsertedNode after the passedNode.
 */
int insertNodeAfter(List passedList, Node passedNode, Node passedInsertedNode) {
	int passedNodeValid = !isNull(passedNode, "Attempting to insert after null Node", FALSE);
	int insertedNodeValid = !isNull(passedNode, "Attempting to insert null Node", FALSE);
	if ((passedNodeValid) && (insertedNodeValid)) {
		// Set existing links
		Node next = passedNode->nextNode;
		if (next != NULL) {
			next->previousNode = passedInsertedNode;
		}
		passedNode->nextNode = passedInsertedNode;

		// Set new links
		passedInsertedNode->previousNode = passedNode;
		passedInsertedNode->nextNode = next;

		// Set new back if applicable
		if (passedList->nodeBack == passedNode) {
			passedList->nodeBack = passedInsertedNode;
		}
		return TRUE;
	}
	return FALSE;
}

/**
 * Checks whether the passed pointer is null, and prints the passed message in that case.
 *
 * If passedIsTerminal is specifically 1, calls exitBadWithMessage with an error message.
 */
int inline isNull(void* passedPointer, char* passedCharArray, int passedIsTerminal) {
	if (passedPointer == NULL) {
		puts(passedCharArray);
		if (passedIsTerminal == TRUE) {
			exitBadWithMessage("Error: Null pointer received.");
		}
		return TRUE;
	}
	return FALSE;
}

/**
 * Checks whether the passed List is empty. Does not perform null check on the passed List.
 */
int inline isListEmpty(List passedList, char* passedCharArray, int passedIsTerminal) {
	if (length(passedList) <= 0) {
		puts(passedCharArray);
		if (passedIsTerminal == TRUE) {
			exitBadWithMessage("Error: List is empty.");
		}
		return TRUE;
	}
	return FALSE;
}

/**
 * Validates whether the passedList's cursor is within a valid range (c > 0) && (c < length)
 */
int inline validateCursor(List passedList) {
	int discoveredIndex = passedList->cursorIndex;
	if ((discoveredIndex < 0) || (discoveredIndex >= passedList->length)) {
		passedList->cursorIndex = UNDEFINED;
	}
	return (discoveredIndex > 0);
}

/**
 * Checks whether the passed List's cursor is valid. Does not perform null check on the passed List.
 */
int inline isCursorValid(List passedList, char* passedCharArray, int passedIsTerminal) {
	if (!validateCursor(passedList)) {
		puts(passedCharArray);
		if (passedIsTerminal == TRUE) {
			exitBadWithMessage("Error: Cursor is invalid.");
		}
		return FALSE;
	}
	return (passedList->nodeCursor != NULL);
}

/**
 * Exits with a code of 1 after printing the passed message.
 */
void inline exitBadWithMessage(const char* passedCharArray) {
	puts(passedCharArray);
	exit(1);
}
